src/Kokkos_HalfPrecision.hpp

```diff
@@ -0,0 +1,173 @@
+// SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+//
+// Questions? Contact Siva Rajamanickam (srajama@sandia.gov)
+//
+// ************************************************************************
+//@HEADER
+*/
+
+#ifndef KOKKOS_HALFPRECISION_HPP
+#define KOKKOS_HALFPRECISION_HPP
+
+#if defined(KOKKOS_ENABLE_CUDA) && defined(HAVE_KOKKOSKERNELS_CUDA_FP16)
+#include <cuda_fp16.h>
+#endif
+
+namespace KokkosKernels {
+    namespace Experimental {
+        /**
+         * Below we check whether the given toolchain has support for portable IEEE-754
+         * FP16 (binary16) precision types. The checks are done via CMake which passes the
+         * results via a KOKKOSKERNELS_HAVE define to KokkosKernels_config.h
+         * 
+         * First we check for cuda half precision support   (HAVE_KOKKOSKERNELS_CUDA_FP16).
+         * Second we check for host half precision support  (HAVE_KOKKOSKERNELS_FP16).
+         * Lastly, we fall back to single precision support.
+         * 
+         * NOTE: If both cuda and host support half precision, the half type will
+         * default to device_fp16_t.
+         */
+        #if defined(KOKKOS_ENABLE_CUDA) && defined(HAVE_KOKKOSKERNELS_CUDA_FP16)
+            using device_fp16_t = __half;
+            using half = device_fp16_t;
+            #if defined(HAVE_KOKKOSKERNELS_FP16)
+                using host_fp16_t = _Float16;
+            #else
+                using host_fp16_t = float;
+            #endif // defined(HAVE_KOKKOSKERNELS_FP16)
+            static KOKKOS_FORCEINLINE_FUNCTION float __cast2float(device_fp16_t x) { return __half2float(x); }
+            static KOKKOS_FORCEINLINE_FUNCTION device_fp16_t __cast2half(float x) { return __float2half(x); }
+            
+        #else // defined(KOKKOS_ENABLE_CUDA) && defined(HAVE_KOKKOSKERNELS_CUDA_FP16)
+            #if defined(HAVE_KOKKOSKERNELS_FP16)
+                using host_fp16_t = _Float16;
+                using half = host_fp16_t;
+            #else // defined(HAVE_KOKKOSKERNELS_FP16)
+                using host_fp16_t = float;
+                using device_fp16_t = host_fp16_t;
+                using half = host_fp16_t;
+            #endif // _Float16
+            static inline float __cast2float(host_fp16_t x) { return (float) x; }
+            static inline host_fp16_t __cast2half(float x) { return (host_fp16_t) x; }
+        #endif
+        ////////////// BEGIN half2float and float2half overloads //////////////
+        /**
+        * Since kokkos does not have support for half precision types yet, we 
+        * must cast to/from float in some kokkos-kernels routines. Except for
+        * the overloads below that actually cast to/from half precision types,
+        * the others should be optimized away by the compiler.
+        */
+        // host_fp16_t
+#if defined(HAVE_KOKKOSKERNELS_FP16)
+        static inline
+        float half2float(host_fp16_t x, float &ret) {
+            ret = __cast2float(x);
```

Comment:
Where is this overload supposed to come from if HAVE_KOKKOSKERNELS_CUDA_FP16 is true?

Reply:
the overload is defined only here if HAVE_KOKKOSKERNELS_FP16 is true. so if both HAVE_KOKKOSKERNELS_FP16 and HAVE_KOKKOSKERNELS_CUDA_FP16 are true, the overload will come from here and `half` will default to device_fp16_t.